### ISSUE TITLE
Can't use `context` predicate with `ContextFound` subscriber

### ISSUE THREAD

**Issue by msabramo** [Pylons/pyramid#1544]
It seems like this should be possible. I'd like to stuff like:

```python
@subscriber(ContextFound, context=Role)
def add_context(event):
    print('*** ContextFound: %r ***' % event.request.context)
```

but I get:

```pytb
  File "/Users/marca/dev/git-repos/pyramid/pyramid/config/__init__.py", line 637, in commit
    self.action_state.execute_actions(introspector=self.introspector)
  File "/Users/marca/dev/git-repos/pyramid/pyramid/config/__init__.py", line 1096, in execute_actions
    tb)
  File "/Users/marca/dev/git-repos/pyramid/pyramid/config/__init__.py", line 1088, in execute_actions
    callable(*args, **kw)
  File "/Users/marca/dev/git-repos/pyramid/pyramid/config/adapters.py", line 55, in register
    order, preds, phash = predlist.make(self, **predicates)
  File "/Users/marca/dev/git-repos/pyramid/pyramid/config/util.py", line 156, in make
    raise ConfigurationError('Unknown predicate values: %r' % (kw,))
pyramid.exceptions.ConfigurationExecutionError: <class 'pyramid.exceptions.ConfigurationError'>: Unknown predicate values: {'context': <class 'kotti_gabriel.resources.Role'>}
  in:
  Line 80 of file /Users/marca/dev/git-repos/pyramid/pyramid/events.py:
    config.add_subscriber(wrapped, iface, **self.predicates)
```

**Comment by msabramo** [comment]
Ah, I just noticed the following at http://docs.pylonsproject.org/docs/pyramid/en/latest/narr/hooks.html#subscriber-predicates:

> There are no default subscriber predicates. You must register one to use one.

**Comment by mmerickel** [comment]
Yeah unfortunately the event system does not really have a standard api for the format of an event so it's difficult to write generic predicates. We could ship default ones but the request is not required to be on an arbitrary event object. So at this point it's left up to you to define any predicates you want to buy into and presumably they will only work on event types you wrote them for.

I think it'd be nice if the API allowed you to register a predicate for a particular type of event instead of allowing it to work with all events but atm that's how it is.

**Comment by msabramo** [comment]
I wonder why there are no default subscriber predicates -- was this a conscious decision to omit them? Perhaps because not all predicates make sense for all events? In that case, maybe they should be implemented but be caveat emptor, or even better, documented.

**Comment by msabramo** [comment]
@mmerickel: See PR #1545.

**Comment by merwok** [comment]
@mmerickel the PR was closed and #2537 opened, does it make this ticket obsolete?

**Comment by mmerickel** [comment]
@merwok both tickets need to be digested .. there's a couple fundamental issues that make improving things difficult.

1. The api for emitting events is not aware of the request object. It is `request.registry.notify()`, not `request.notify()`.
2. There is no standard api for events and so accessing the request/context from an event cannot be done generically.

#2537 is an approach that tries to solve the problem without adding a new notify api (it's focused on solving problem 2 above) - but it also makes it less generic. Alternatively, if we solved option 1 -- if events were emitted through the request object then it'd be way more straightforward (possibly even trivial) to register default predicates for subscribers.

**Comment by mmerickel** [comment]
To be clear it’s mostly a compatibility question but if I was designing subscriber predicates today I would have done it like this: define `request.notify(event)` and have the predicate api be `(event, request)` similar to route predicates that are `(info, request)` and view predicates that are `(context, request)`. 

Maybe there’s a way to get there from here without breaking compat if someone wanted to dive in. The more straightforward way to support things that preserves compat is https://github.com/Pylons/pyramid/issues/2537

**Comment by merwok** [comment]
> Maybe there’s a way to get there from here without breaking compat if someone wanted to dive in.

Something like using `functool.signature` to see what arguments are expected by a predicate?
